src/session_impl.cpp

```diff
@@ -283,6 +278,66 @@ namespace aux {
 		}
 	}
 
+	void expand_devices(span<ip_interface const> const ifs
+		, span<ip_route const> const routes
+		, std::vector<listen_endpoint_t>& eps)
+	{
+		for (auto& ep : eps)
+		{
+			auto const iface = ep.device.empty()
+				? std::find_if(ifs.begin(), ifs.end(), [&](ip_interface const& ipface)
+					{
+						return match_addr_mask(ipface.interface_address, ep.addr, ipface.netmask);
+					})
+				: std::find_if(ifs.begin(), ifs.end(), [&](ip_interface const& ipface)
+					{
+						return ipface.name == ep.device
+							&& match_addr_mask(ipface.interface_address, ep.addr, ipface.netmask);
+					});
+
+			if (iface == ifs.end())
+			{
+				// we can't find which device this is for, just assume we can't
+				// reach anything on it
+				ep.netmask = build_netmask(0, ep.addr.is_v4() ? AF_INET : AF_INET6);
+				continue;
+			}
+
+			ep.netmask = iface->netmask;
+
+			// also record whether the device has a gateway associated with it
+			// (which indicates it can be used to reach the internet)
+			bool const has_gateway
+				= std::find_if(routes.begin(), routes.end(), [&](ip_route const& r)
+				{
+					return r.destination.is_unspecified()
+						&& r.destination.is_v4() == iface->interface_address.is_v4()
+						&& strcmp(r.name, iface->name) == 0;
+				}) != routes.end();
```

Comment:
`has_gateway`should not be set to true for listen endpoints on local IPv6 addresses